src/bootstrap/check.rs

```diff
@@ -115,7 +107,37 @@ pub fn cargo(build: &Build, stage: u32, host: &str) {
     // available.
     cargo.env("CFG_DISABLE_CROSS_TESTS", "1");
 
-    build.run(cargo.env("PATH", newpath));
+    build.run(cargo.env("PATH", &path_for_cargo(build, compiler)));
+}
+
+/// Runs `cargo test` for the rls.
+pub fn rls(build: &Build, stage: u32, host: &str) {
+    let compiler = &Compiler::new(stage, host);
+
+    let mut cargo = build.cargo(compiler, Mode::Tool, host, "test");
+    cargo.arg("--manifest-path").arg(build.src.join("src/tools/rls/Cargo.toml"));
+
+    // Don't build tests dynamically, just a pain to work with
+    cargo.env("RUSTC_NO_PREFER_DYNAMIC", "1");
+
+    let mut dylib_path = dylib_path();
+    dylib_path.insert(0, build.sysroot_libdir(&compiler, host));
+    cargo.env(dylib_path_var(), env::join_paths(&dylib_path).unwrap());
+
+    // Avoid using the rustc shim.
+    cargo.env("RUSTC", build.compiler_path(compiler));
+
+    build.run(cargo.env("PATH", &path_for_cargo(build, compiler)));
```

Comment:
Does the rls work the same way as cargo in the sense that it's matching output? Cargo will print the value of `RUSTC` at some point during testing which the testing itself didn't expect, so Cargo's tests are basically asserting that `RUSTC=rustc` so that's why this was necessary. Ideally though this modification of `PATH` wouldn't be necessary, but do the rls tests do basically the same thing?